src/Interpreters/Access/InterpreterGrantQuery.cpp

```diff
@@ -330,6 +331,67 @@ namespace
             updateGrantedAccessRightsAndRolesTemplate(*role, elements_to_grant, elements_to_revoke, roles_to_grant, roles_to_revoke, admin_option);
     }
 
+    template <typename T>
+    void grantCurrentGrantsTemplate(
+        T & grantee,
+        const AccessRightsElements & elements_to_grant,
+        const AccessRightsElements & elements_to_revoke,
+        std::shared_ptr<const ContextAccess> current_user_access,
+        bool grant_option)
+    {
+        if (!elements_to_revoke.empty())
+            grantee.access.revoke(elements_to_revoke);
+
+        /// We need to collect all current user's grant and filter them by grant option.
+        AccessRightsElements current_user_grantable_elements;
+        auto available_grant_elements = current_user_access->getAccessRights()->getElements();
+        std::copy_if(
+            available_grant_elements.begin(),
+            available_grant_elements.end(),
+            std::back_inserter(current_user_grantable_elements),
+            [](AccessRightsElement x) { return x.grant_option || x.is_partial_revoke; });
+
+        AccessRights current_user_rights;
+        for (auto & element : current_user_grantable_elements)
+        {
+            if (element.is_partial_revoke)
+                current_user_rights.revoke(element);
+            else
+                current_user_rights.grant(element);
+        }
+
+        /// If elements_to_grant was not specified it will grant all available for user grants.
+        /// Otherwise, we will intersect available grants with given set.
+        if (elements_to_grant.empty())
```

Comment:
What about `GRANT CURRENT GRANTS(NONE ON *.*)`? I know this is a strange statement, but it's still valid.

Reply:
And this case is worth a test too.

Reply:
https://github.com/ClickHouse/ClickHouse/pull/48262/files#diff-9d106b2999421e27ea1924830817d4b8a6ea9fc093f797fd30f9a592fe9f338eR616